modules/ethereum-contract/builtin/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use bridge_node_runtime::{BlockNumber, Hash, Header as RuntimeHeader};
+use codec::{Decode, Encode};
+use ethereum_types::U256;
+use sp_blockchain::Error as ClientError;
+use sp_finality_grandpa::{AuthorityList, ConsensusLog, GRANDPA_ENGINE_ID};
+
+/// Builtin errors.
+#[derive(Debug)]
+pub enum Error {
+	/// Failed to decode block number.
+	BlockNumberDecode,
+	/// Failed to decode Substrate header.
+	HeaderDecode(codec::Error),
+	/// Failed to decode best voters set.
+	BestSetDecode(codec::Error),
+	/// Failed to decode finality proof.
+	FinalityProofDecode(codec::Error),
+	/// Failed to verify justification.
+	JustificationVerify(ClientError),
+}
+
+/// Substrate header.
+#[derive(Debug)]
+pub struct Header {
+	/// Header hash.
+	pub hash: Hash,
+	/// Parent header hash.
+	pub parent_hash: Hash,
+	/// Header number.
+	pub number: BlockNumber,
+	/// GRANDPA validators change signal.
+	pub signal: Option<ValidatorsSetSignal>,
+}
+
+/// GRANDPA validators set change signal.
+#[derive(Debug)]
+pub struct ValidatorsSetSignal {
+	/// Signal delay.
+	pub delay: BlockNumber,
+	/// New validators set.
+	pub validators: Vec<u8>,
+}
+
+/// Convert from U256 to BlockNumber.
+pub fn to_substrate_block_number(number: U256) -> Result<BlockNumber, Error> {
+	match number == number.low_u32().into() {
+		true => Ok(number.low_u32()),
+		false => Err(Error::BlockNumberDecode),
+	}
+}
+
+/// Convert from BlockNumber to U256.
+pub fn from_substrate_block_number(number: BlockNumber) -> Result<U256, Error> {
+	Ok(U256::from(number as u64))
+}
+
+/// Parse Substrate header.
+pub fn parse_substrate_header(raw_header: &[u8]) -> Result<Header, Error> {
+	RuntimeHeader::decode(&mut &raw_header[..])
+		.map(|header| Header {
+			hash: header.hash(),
+			parent_hash: header.parent_hash,
+			number: header.number,
+			signal: sp_runtime::traits::Header::digest(&header)
+				.log(|log| {
+					log.as_consensus().and_then(|(engine_id, log)| {
+						if engine_id == GRANDPA_ENGINE_ID {
+							Some(log)
+						} else {
+							None
+						}
+					})
+				})
+				.and_then(|log| ConsensusLog::decode(&mut &log[..]).ok())
+				.and_then(|log| match log {
+					ConsensusLog::ScheduledChange(scheduled_change) => Some(ValidatorsSetSignal {
+						delay: scheduled_change.delay,
+						validators: scheduled_change.next_authorities.encode(),
+					}),
+					_ => None,
+				}),
+		})
+		.map_err(Error::HeaderDecode)
+}
+
+/// Verify GRANDPA finality proof.
+pub fn verify_substrate_finality_proof(
+	finality_target_number: BlockNumber,
+	finality_target_hash: Hash,
+	best_set_id: u64,
+	raw_best_set: &[u8],
+	raw_finality_proof: &[u8],
```

Comment:
@HCastano @tomusdrw TIL (not actually today, but...) that we may listen to `Commit` messages on full nodes && finalize headers using these messages. Let me refresh your memory - `Justifications` are **rare** and if we'd depend on justifications, then we may only finalize headers in huge bunches (like ~30000 blocks - see https://github.com/paritytech/substrate/issues/4728 for example). That is the problem for native light client as well.

We may also want to support finalization by `Commit` messages here. The problem is that (afaiu) these messages are not persistent && we can't access these from RPCs. So if we would choose to use it, then we need some RPCs to deal with that. I'm not sure if that's a part of that finality subscription RPC, Hernando is working on? Wdyt - do we need it now, or it could be added later?

Reply:
RPC that @HCastano is working on, will emit non-persistent justifications, made of intermediate commits. So the only thing we need is to use that RPC in relay, when it'll be ready.

Reply:
logged: #71